include/boost/geometry/iterators/detail/segment_iterator/range_segment_iterator.hpp

```diff
@@ -100,23 +100,40 @@ class range_segment_iterator
             Reference
         >
 {
+private:
+    static bool has_less_than_two_elements(Range const& r)
+    {
+        iterator_type first = range_iterator_begin<Range>::apply(r);
+        iterator_type last = range_iterator_end<Range>::apply(r);
+
+        return first == last || ++first == last;
```

Comment:
If the range is a random access range, calling `std::distance` is perfectly okay.
If the range is not a random access range, calling `std::distance` takes linear time in the size of the range.

It is true that in our context, the range passed here is either a ring or a linestring, and our corresponding concepts require that they are of random access. So it is okay to use the following return statement:
```
return std::distance(first, last) < 3u;
```
Let me know if this preferable, and I will make the change.